File: backend/app/laps/extrema.py
"""Lap-vs-lap comparison: ported from legacy/overlay/lap_timer_v7.py:147-293.

The original rendered matplotlib PNGs from an interactive `input()` CLI
loop. Here the same math (local speed maxima/minima = braking/corner points,
straight-line speeds) is kept, but it returns plain data structures — the
webapp renders this as an interactive chart in the browser instead of a
static image.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np
import pandas as pd


@dataclass
class LapComparison:
    lap_a: int
    lap_b: int
    duration_a: float
    duration_b: float
    series_a: pd.DataFrame  # columns: rel_time, speed
    series_b: pd.DataFrame
    maxima_a: list[tuple[float, float]]  # (rel_time, speed)
    minima_a: list[tuple[float, float]]
    maxima_b: list[tuple[float, float]]
    minima_b: list[tuple[float, float]]


def find_local_extrema(values: pd.Series | list[float], window: int = 30, mode: str = "max") -> list[tuple[int, float]]:
    """Returns (index, value) for points that are the local max/min within
    +/- `window` samples on both sides — i.e. corner apexes (min) or braking
    points / straight-line peaks (max), not just noisy single-sample spikes.
    """
    vals = list(values)
    extrema = []
    for i in range(window, len(vals) - window):
        center = vals[i]
        neighborhood = vals[i - window:i] + vals[i + 1:i + window + 1]
        if mode == "max" and all(center >= x for x in neighborhood):
            extrema.append((i, center))
        elif mode == "min" and all(center <= x for x in neighborhood):
            extrema.append((i, center))
    return extrema


def _lap_slice(df: pd.DataFrame, lap_indices: list[int], lap_number: int) -> pd.DataFrame:
    s_idx, e_idx = lap_indices[lap_number - 1], lap_indices[lap_number]
    sl = df.iloc[s_idx:e_idx].copy()
    sl["rel_time"] = sl["time"] - sl.iloc[0]["time"]
    return sl


def compare_laps(
    df: pd.DataFrame,
    lap_table: pd.DataFrame,
    lap_indices: list[int],
    lap_a: int,
    lap_b: int,
    extrema_window: int = 30,
) -> LapComparison:
    n_laps = len(lap_indices) - 1
    if not (1 <= lap_a <= n_laps) or not (1 <= lap_b <= n_laps):
        raise ValueError(f"Lap numbers must be within 1..{n_laps}")

    slice_a = _lap_slice(df, lap_indices, lap_a)
    slice_b = _lap_slice(df, lap_indices, lap_b)

    max_a = find_local_extrema(slice_a["speed"], window=extrema_window, mode="max")
    min_a = find_local_extrema(slice_a["speed"], window=extrema_window, mode="min")
    max_b = find_local_extrema(slice_b["speed"], window=extrema_window, mode="max")
    min_b = find_local_extrema(slice_b["speed"], window=extrema_window, mode="min")

    t_a = slice_a["rel_time"].to_numpy()
    t_b = slice_b["rel_time"].to_numpy()

    def _to_points(extrema, t):
        return [(float(t[i]), float(v)) for i, v in extrema]

    dur_a = float(lap_table.loc[lap_table["lap"] == lap_a, "duration"].iloc[0])
    dur_b = float(lap_table.loc[lap_table["lap"] == lap_b, "duration"].iloc[0])

    return LapComparison(
        lap_a=lap_a,
        lap_b=lap_b,
        duration_a=dur_a,
        duration_b=dur_b,
        series_a=slice_a[["rel_time", "speed"]].reset_index(drop=True),
        series_b=slice_b[["rel_time", "speed"]].reset_index(drop=True),
        maxima_a=_to_points(max_a, t_a),
        minima_a=_to_points(min_a, t_a),
        maxima_b=_to_points(max_b, t_b),
        minima_b=_to_points(min_b, t_b),
    )

File: frontend/src/lib/laps/extrema.ts
/**
 * Lap-vs-lap comparison: ported from backend/app/laps/extrema.py.
 */

import type { LapTableRow, TelemetryRow } from './detection'

// Field names mirror LapComparison's Python dataclass columns verbatim
// (like detection.ts's TelemetryRow/LapTableRow) rather than camelCase --
// this is data shape, not an API surface, and staying byte-for-byte
// comparable to extrema.py's fields keeps this port easy to audit.
export interface LapComparison {
  lap_a: number
  lap_b: number
  duration_a: number
  duration_b: number
  series_a: { rel_time: number; speed: number }[]
  series_b: { rel_time: number; speed: number }[]
  maxima_a: [number, number][]
  minima_a: [number, number][]
  maxima_b: [number, number][]
  minima_b: [number, number][]
}

/** Returns [index, value] for points that are the local max/min within +/-
 * `window` samples on both sides -- corner apexes (min) or braking points /
 * straight-line peaks (max), not just noisy single-sample spikes. */
export function findLocalExtrema(
  values: number[],
  window = 30,
  mode: 'max' | 'min' = 'max',
): [number, number][] {
  const extrema: [number, number][] = []
  for (let i = window; i < values.length - window; i++) {
    const center = values[i]
    const neighborhood = [...values.slice(i - window, i), ...values.slice(i + 1, i + window + 1)]
    if (mode === 'max' && neighborhood.every((x) => center >= x)) extrema.push([i, center])
    else if (mode === 'min' && neighborhood.every((x) => center <= x)) extrema.push([i, center])
  }
  return extrema
}

function lapSlice(rows: TelemetryRow[], lapIndices: number[], lapNumber: number): (TelemetryRow & { rel_time: number })[] {
  const sIdx = lapIndices[lapNumber - 1]
  const eIdx = lapIndices[lapNumber]
  const slice = rows.slice(sIdx, eIdx)
  const t0 = slice.length ? slice[0].time : 0
  return slice.map((r) => ({ ...r, rel_time: r.time - t0 }))
}

export function compareLaps(
  rows: TelemetryRow[],
  lapTable: LapTableRow[],
  lapIndices: number[],
  lapA: number,
  lapB: number,
  extremaWindow = 30,
): LapComparison {
  const nLaps = lapIndices.length - 1
  if (!(lapA >= 1 && lapA <= nLaps) || !(lapB >= 1 && lapB <= nLaps)) {
    throw new Error(`Lap numbers must be within 1..${nLaps}`)
  }

  const sliceA = lapSlice(rows, lapIndices, lapA)
  const sliceB = lapSlice(rows, lapIndices, lapB)

  const speedA = sliceA.map((r) => r.speed)
  const speedB = sliceB.map((r) => r.speed)
  const maxA = findLocalExtrema(speedA, extremaWindow, 'max')
  const minA = findLocalExtrema(speedA, extremaWindow, 'min')
  const maxB = findLocalExtrema(speedB, extremaWindow, 'max')
  const minB = findLocalExtrema(speedB, extremaWindow, 'min')

  const toPoints = (extrema: [number, number][], slice: (TelemetryRow & { rel_time: number })[]): [number, number][] =>
    extrema.map(([i, v]) => [slice[i].rel_time, v])

  const durationA = lapTable.find((l) => l.lap === lapA)!.duration
  const durationB = lapTable.find((l) => l.lap === lapB)!.duration

  return {
    lap_a: lapA,
    lap_b: lapB,
    duration_a: durationA,
    duration_b: durationB,
    series_a: sliceA.map((r) => ({ rel_time: r.rel_time, speed: r.speed })),
    series_b: sliceB.map((r) => ({ rel_time: r.rel_time, speed: r.speed })),
    maxima_a: toPoints(maxA, sliceA),
    minima_a: toPoints(minA, sliceA),
    maxima_b: toPoints(maxB, sliceB),
    minima_b: toPoints(minB, sliceB),
  }
}
